### ISSUE TITLE
`local_add_canonizer` and `local_mul_canonizer` fail to merge all nested operations of the same kind

### ISSUE THREAD

**Issue by ricardoV94** [aesara-devs/aesara#1244]
Sometimes the `local_add_canonizer` and `local_mul_canonizer` fail to merge all nested cases. The pattern I notice is when we combine them with the inverse Op:

```python
import aesara
import aesara.tensor as at
from aesara.compile.mode import Mode

x, y, z = at.vectors("x", "y", "z")

mul = x * y * z
add = x + y + z

out = mul - add
f = aesara.function([x, y, z], [out], mode=Mode().including("canonicalize").excluding("fusion"))
aesara.dprint(f)
```
```
Elemwise{Sub}[(0, 0)] [id A] 3
 |Elemwise{mul,no_inplace} [id B] 2
 | |x [id C]
 | |y [id D]
 | |z [id E]
 |Elemwise{Add}[(0, 0)] [id F] 1
   |Elemwise{add,no_inplace} [id G] 0
   | |x [id C]
   | |y [id D]
   |z [id E]
```
Or, conversely
```python
out = add / mul
f = aesara.function([x, y, z], [out], mode=Mode().including("canonicalize").excluding("fusion"))
aesara.dprint(f)
```
```
Elemwise{TrueDiv}[(0, 0)] [id A] 3
 |Elemwise{add,no_inplace} [id B] 2
 | |x [id C]
 | |y [id D]
 | |z [id E]
 |Elemwise{Mul}[(0, 0)] [id F] 1
   |Elemwise{mul,no_inplace} [id G] 0
   | |x [id C]
   | |y [id D]
   |z [id E]
Out[5]: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
```

_Originally posted by @ricardoV94 in https://github.com/aesara-devs/aesara/issues/1243#issuecomment-1271800357_

**Comment by brandonwillard** [comment]
It would help to set `optimizer_verbose=True`; that way we can see the types and order of the rewrites that were applied.

Anyway, I looked at the logic in `AlgebraicCanonizer.transform`, and it appears as thought [the `out_clients` check](https://github.com/aesara-devs/aesara/blob/a619a4eca5c9207c8059533230cebd0f534e29b2/aesara/tensor/rewriting/math.py#L995) in the beginning of that method is finding the `sub` client (i.e. `id A` in the first example) and aborting a rewrite of the first `add` (i.e. `id F`).

I haven't thought about it too carefully, but it does seem like that first `add` node should be "flattened" (i.e. turned into `add(x, y, z)`) regardless of the clients.  Nevertheless, the rewrite seems to want the outer-most `sub` (i.e. `id A`) to be merged with the `add` (i.e. `id F`)&mdash;or something similar&mdash;first.

The biggest problem here is that `AlgebraicCanonizer` is a mess.  It's not entirely clear which rewrites it's suppose to represent and/or what it's concerns and constraints are.  For instance, why can't we have a simple stand-alone node rewrite that consistently flattens **all** variadic binary `Op`s like these?  Was the intention to avoid making bold choices about the order in which operations take place (e.g. floating point concerns).  If so, why does `AlgebraicCanonizer` flatten anything at all?

Any amount of work that clears up these ambiguities and apparent inconsistencies of `AlgebraicCanonizer` would go a long way.  Ideally, we would be able to replace `AlgebraicCanonizer` with a small set of simple, self-contained node rewrites&mdash;and I'm pretty sure that it can be done, even if there are some important restrictions limiting when such rewrites should be applied.

N.B. There are a few related rewrite issues: e.g. https://github.com/aesara-devs/aesara/issues/287, https://github.com/aesara-devs/aesara/issues/275.

**Comment by ricardoV94** [comment]
>  For instance, why can't we have a simple stand-alone node rewrite that consistently flattens all variadic binary Ops like these? 

We do already, it's the one I was considering removing in #1243 because it seems like the canonizers were trying to do the same thing anyway.

**Comment by brandonwillard** [comment]
> > For instance, why can't we have a simple stand-alone node rewrite that consistently flattens all variadic binary Ops like these?
> 
> We do already, it's the one I was considering removing in #1243 because it seems like the canonizers were trying to do the same thing anyway.

Yes, the redundancy is clearly not needed, but as I mentioned above, the `AlgebraicCanonizer` class is a monolithic mess, and that's not the direction in which we want to continue going.  If we can break _all_ of it down to independent rewrites like that, then we would be much better off, especially since the redundancies would be much easier to spot, and that's something we truly need.
